File: win_app_2/src/services/translation/deeplLanguage.ts
export interface DeepLLanguageInfo {
  /** DeepL code (passed to API as target_lang). */
  code: string
  displayName: string
  flag: string
  /** Short badge label (2-3 chars). */
  badge: string
}

export const DEEPL_LANGUAGES: DeepLLanguageInfo[] = [
  { code: 'BG', displayName: 'Bulgarian', flag: '🇧🇬', badge: 'BG' },
  { code: 'CS', displayName: 'Czech', flag: '🇨🇿', badge: 'CS' },
  { code: 'DA', displayName: 'Danish', flag: '🇩🇰', badge: 'DA' },
  { code: 'DE', displayName: 'German', flag: '🇩🇪', badge: 'DE' },
  { code: 'EL', displayName: 'Greek', flag: '🇬🇷', badge: 'EL' },
  { code: 'EN-US', displayName: 'English (US)', flag: '🇺🇸', badge: 'EN' },
  { code: 'EN-GB', displayName: 'English (UK)', flag: '🇬🇧', badge: 'EN' },
  { code: 'ES', displayName: 'Spanish', flag: '🇪🇸', badge: 'ES' },
  { code: 'ET', displayName: 'Estonian', flag: '🇪🇪', badge: 'ET' },
  { code: 'FI', displayName: 'Finnish', flag: '🇫🇮', badge: 'FI' },
  { code: 'FR', displayName: 'French', flag: '🇫🇷', badge: 'FR' },
  { code: 'HU', displayName: 'Hungarian', flag: '🇭🇺', badge: 'HU' },
  { code: 'ID', displayName: 'Indonesian', flag: '🇮🇩', badge: 'ID' },
  { code: 'IT', displayName: 'Italian', flag: '🇮🇹', badge: 'IT' },
  { code: 'JA', displayName: 'Japanese', flag: '🇯🇵', badge: 'JA' },
  { code: 'KO', displayName: 'Korean', flag: '🇰🇷', badge: 'KO' },
  { code: 'LT', displayName: 'Lithuanian', flag: '🇱🇹', badge: 'LT' },
  { code: 'LV', displayName: 'Latvian', flag: '🇱🇻', badge: 'LV' },
  { code: 'NB', displayName: 'Norwegian', flag: '🇳🇴', badge: 'NB' },
  { code: 'NL', displayName: 'Dutch', flag: '🇳🇱', badge: 'NL' },
  { code: 'PL', displayName: 'Polish', flag: '🇵🇱', badge: 'PL' },
  { code: 'PT-BR', displayName: 'Portuguese (Brazil)', flag: '🇧🇷', badge: 'PT' },
  { code: 'PT-PT', displayName: 'Portuguese (Portugal)', flag: '🇵🇹', badge: 'PT' },
  { code: 'RO', displayName: 'Romanian', flag: '🇷🇴', badge: 'RO' },
  { code: 'RU', displayName: 'Russian', flag: '🇷🇺', badge: 'RU' },
  { code: 'SK', displayName: 'Slovak', flag: '🇸🇰', badge: 'SK' },
  { code: 'SL', displayName: 'Slovenian', flag: '🇸🇮', badge: 'SL' },
  { code: 'SV', displayName: 'Swedish', flag: '🇸🇪', badge: 'SV' },
  { code: 'TR', displayName: 'Turkish', flag: '🇹🇷', badge: 'TR' },
  { code: 'UK', displayName: 'Ukrainian', flag: '🇺🇦', badge: 'UK' },
  { code: 'ZH', displayName: 'Chinese', flag: '🇨🇳', badge: 'ZH' },
]

const LOOKUP = new Map(DEEPL_LANGUAGES.map((l) => [l.code, l]))

export function getLanguageInfo(code: string | null | undefined): DeepLLanguageInfo | undefined {
  if (!code) return undefined
  return LOOKUP.get(code.toUpperCase())
}

/**
 * Map a BCP-47 / ISO code to a DeepL language. Returns null if no reasonable
 * mapping exists.
 */
export function fromBCP47(code: string): DeepLLanguageInfo | null {
  const normalized = code.toUpperCase().replace(/_/g, '-')
  const exact = LOOKUP.get(normalized)
  if (exact) return exact
  const primary = normalized.split('-')[0]
  switch (primary) {
    case 'EN':
      return LOOKUP.get('EN-US') ?? null
    case 'PT':
      return LOOKUP.get('PT-PT') ?? null
    case 'ZH':
      return LOOKUP.get('ZH') ?? null
    default:
      return LOOKUP.get(primary) ?? null
  }
}

File: ios_app/QueenMamaIOS/Services/Translation/DeepLLanguage.swift
import Foundation

/// DeepL target languages (subset of DeepL's full language list).
/// Raw values are DeepL's language codes — passed as `target_lang` in API calls.
/// See: https://developers.deepl.com/docs/getting-started/supported-languages
enum DeepLLanguage: String, CaseIterable, Identifiable, Sendable {
    case bulgarian      = "BG"
    case czech          = "CS"
    case danish         = "DA"
    case german         = "DE"
    case greek          = "EL"
    case englishUS      = "EN-US"
    case englishGB      = "EN-GB"
    case spanish        = "ES"
    case estonian       = "ET"
    case finnish        = "FI"
    case french         = "FR"
    case hungarian      = "HU"
    case indonesian     = "ID"
    case italian        = "IT"
    case japanese       = "JA"
    case korean         = "KO"
    case lithuanian     = "LT"
    case latvian        = "LV"
    case norwegianBokmal = "NB"
    case dutch          = "NL"
    case polish         = "PL"
    case portugueseBR   = "PT-BR"
    case portuguesePT   = "PT-PT"
    case romanian       = "RO"
    case russian        = "RU"
    case slovak         = "SK"
    case slovenian      = "SL"
    case swedish        = "SV"
    case turkish        = "TR"
    case ukrainian      = "UK"
    case chinese        = "ZH"

    var id: String { rawValue }

    /// Human-readable name in English (used in pickers).
    var displayName: String {
        switch self {
        case .bulgarian:        return "Bulgarian"
        case .czech:            return "Czech"
        case .danish:           return "Danish"
        case .german:           return "German"
        case .greek:            return "Greek"
        case .englishUS:        return "English (US)"
        case .englishGB:        return "English (UK)"
        case .spanish:          return "Spanish"
        case .estonian:         return "Estonian"
        case .finnish:          return "Finnish"
        case .french:           return "French"
        case .hungarian:        return "Hungarian"
        case .indonesian:       return "Indonesian"
        case .italian:          return "Italian"
        case .japanese:         return "Japanese"
        case .korean:           return "Korean"
        case .lithuanian:       return "Lithuanian"
        case .latvian:          return "Latvian"
        case .norwegianBokmal:  return "Norwegian"
        case .dutch:            return "Dutch"
        case .polish:           return "Polish"
        case .portugueseBR:     return "Portuguese (Brazil)"
        case .portuguesePT:     return "Portuguese (Portugal)"
        case .romanian:         return "Romanian"
        case .russian:          return "Russian"
        case .slovak:           return "Slovak"
        case .slovenian:        return "Slovenian"
        case .swedish:          return "Swedish"
        case .turkish:          return "Turkish"
        case .ukrainian:        return "Ukrainian"
        case .chinese:          return "Chinese"
        }
    }

    /// Flag emoji (best-effort, regional variants share base flag).
    var flag: String {
        switch self {
        case .bulgarian:        return "🇧🇬"
        case .czech:            return "🇨🇿"
        case .danish:           return "🇩🇰"
        case .german:           return "🇩🇪"
        case .greek:            return "🇬🇷"
        case .englishUS:        return "🇺🇸"
        case .englishGB:        return "🇬🇧"
        case .spanish:          return "🇪🇸"
        case .estonian:         return "🇪🇪"
        case .finnish:          return "🇫🇮"
        case .french:           return "🇫🇷"
        case .hungarian:        return "🇭🇺"
        case .indonesian:       return "🇮🇩"
        case .italian:          return "🇮🇹"
        case .japanese:         return "🇯🇵"
        case .korean:           return "🇰🇷"
        case .lithuanian:       return "🇱🇹"
        case .latvian:          return "🇱🇻"
        case .norwegianBokmal:  return "🇳🇴"
        case .dutch:            return "🇳🇱"
        case .polish:           return "🇵🇱"
        case .portugueseBR:     return "🇧🇷"
        case .portuguesePT:     return "🇵🇹"
        case .romanian:         return "🇷🇴"
        case .russian:          return "🇷🇺"
        case .slovak:           return "🇸🇰"
        case .slovenian:        return "🇸🇮"
        case .swedish:          return "🇸🇪"
        case .turkish:          return "🇹🇷"
        case .ukrainian:        return "🇺🇦"
        case .chinese:          return "🇨🇳"
        }
    }

    /// Short badge label (2-3 chars for compact UI badges).
    var badge: String {
        switch self {
        case .englishUS:        return "EN"
        case .englishGB:        return "EN"
        case .portugueseBR:     return "PT"
        case .portuguesePT:     return "PT"
        default:                return rawValue
        }
    }

    /// Maps a BCP-47 / ISO language code (as used in `ConfigurationManager.primaryLanguage`)
    /// to a DeepL language. Returns nil if no reasonable mapping exists.
    static func fromBCP47(_ code: String) -> DeepLLanguage? {
        let normalized = code.uppercased().replacingOccurrences(of: "_", with: "-")
        // Exact match first
        if let exact = DeepLLanguage(rawValue: normalized) {
            return exact
        }
        // Strip region for two-letter primary codes
        let primary = String(normalized.split(separator: "-").first ?? "")
        switch primary {
        case "EN": return .englishUS
        case "PT": return .portuguesePT
        case "ZH": return .chinese
        default:
            return DeepLLanguage(rawValue: primary)
        }
    }
}
